lib/cli/src/commands/run/wasi.rs

```diff
@@ -225,21 +249,161 @@ impl Wasi {
         Ok((wasi_env, instance))
     }
 
+    // Runs the Wasi process
+    pub fn run(run: RunProperties, store: Store) -> Result<i32> {
+        let tasks = run.ctx.data(&store).tasks().clone();
+
+        // The return value is passed synchronously and will block until the result is returned
+        // this is because the main thread can go into a deep sleep and exit the dedicated thread
+        let (tx, rx) = std::sync::mpsc::channel();
+
+        // We run it in a blocking thread as the WASM function may otherwise hold
+        // up the IO operations
+        tasks.task_dedicated(Box::new(move || {
+            Self::run_with_deep_sleep(run, store, tx, None);
```

Comment:
This should only be needed when the functionality is enabled, but:

I don't see the point of supporting sleep for the main thread in the CLI. 
What's the motivation?
Just for testing?
Child threads can make sense when opted into, but otherwise this just seems to add code comlexity.

Reply:
the motivation was that asynchronous threading is higher performance (when under high number of threads) as shown with some performance test runs with minimal extra overhead therefore the default of "on" was possibly the preferred for option for consumers however I get your point that reduced capabilities may be the preferred as long as they can opt-in.

i've patched it and switched from opt-out to opt-in

Reply:
thinking about this some more there may be a very good reason to do this....

`tiered compilation`

asynchronous threads allow a safe way to unload and reload a new module, hence it may become fairly trivial to build on top of this patch and introduce a mechanism where it defaults to `single-pass` and then upgrades to `cranelift` on the next rewind

Reply:
Slight misunderstanding there: this wasn't a general remark, but just specifically about support deep sleep for the main thread in the CLI - which I don't think is necessary.

Reply:
its need in the CLI to support tiered compilation from the CLI as otherwise the main thread will not be able to reload the module after its upgraded from `singlepass` to `cranelift`